src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1107,29 +1107,40 @@ impl Cage {
             None => STARTINGFD,
         };
 
+        if start_fd == fd { return start_fd; } //if the file descriptors are equal (which is not plausible here), return
+
+        // get the filedesc_enum
+        let checkedfd = self.get_filedescriptor(fd).unwrap();
+        let filedesc_enum = checkedfd.write();
+        let filedesc_enum = if let Some(f) = &*filedesc_enum {f} else {
+            return syscall_error(Errno::EBADF, "dup2","Invalid old file descriptor.");
+        };
+
         //checking whether the fd exists in the file table
-        return Self::_dup2_helper(&self, fd, start_fd, false)
+        return Self::_dup2_helper(&self, filedesc_enum, start_fd, false)
     }
 
     pub fn dup2_syscall(&self, oldfd: i32, newfd: i32) -> i32{
-        //if the old fd exists, execute the helper, else return error
-        return Self::_dup2_helper(&self, oldfd, newfd, true);
-    }
-
-    pub fn _dup2_helper(&self, oldfd: i32, newfd: i32, fromdup2: bool) -> i32 {
         //checking if the new fd is out of range
         if newfd >= MAXFD || newfd < 0 {
-            return syscall_error(Errno::EBADF, "dup or dup2", "provided file descriptor is out of range");
-        }
+           return syscall_error(Errno::EBADF, "dup or dup2", "provided file descriptor is out of range");
```

Comment:
`return syscall_error(Errno::EBADF, "dup2", "provided file descriptor is out of range")`

Reply:
That's an interesting catch. Its outside of the scope of this PR a bit but does pertain to the same material.

@coulson Can you test removing the close_result lines so the close is silent?

Reply:
the close_result lines should not be removed, because if the close fail, dup2 should also return error, right?

Reply:
The man page says that it doesn't

Reply:
Well, I checked Linux code, dup2 does return normally no matter the close succeed or not. I will change this code to be align with this.